lib/datalog/access.go

```diff
@@ -0,0 +1,366 @@
+type NodeAccessRequest struct {
+	Username  string
+	Login     string
+	Node      string
+	Namespace string
+}
+
+// NodeAccessResponse defines all interpreted facts from the input facts.
+type NodeAccessResponse struct {
+	Accesses        Facts
+	facts           Facts
+	mappings        map[string]uint32
+	reverseMappings map[uint32]string
+	errorStrings    map[int]string
+}
+
+const (
+	loginTraitHash = 0
+	keyJoin        = "_"
+	userIndex      = 0
+	loginIndex     = 1
+	nodeIndex      = 2
+	roleIndex      = 3
+)
+
+// QueryNodeAccess returns a list of accesses to Teleport.
+func QueryNodeAccess(ctx context.Context, client auth.ClientI, req NodeAccessRequest) (*NodeAccessResponse, error) {
+	resp := NewNodeAccessResponse()
+	resp.addToMap(types.Wildcard)
+
+	// Since we can filter on specific usernames, this conditional will only retrieve and build the facts for the required user(s).
+	if req.Username != "" {
+		u, err := client.GetUser(req.Username, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resp.createUserPredicates(u, req.Login)
+	} else {
+		us, err := client.GetUsers(false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, u := range us {
+			resp.createUserPredicates(u, req.Login)
+		}
+	}
+
+	rs, err := client.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, r := range rs {
+		resp.createRolePredicates(r)
+	}
+
+	ns, err := client.GetNodes(ctx, req.Namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, n := range ns {
+		if len(req.Node) == 0 || n.GetHostname() == req.Node {
+			resp.createNodePredicates(n)
+		}
+	}
+
+	b, err := resp.facts.Marshal()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ptr := (*C.uchar)(C.CBytes(b))
+	defer C.free(unsafe.Pointer(ptr))
+	res := (*C.uchar)(C.CBytes(make([]byte, len(b))))
+	l := C.process_access(ptr, res, C.size_t(len(b)), C.size_t(len(b)))
+	if int(l) < 0 {
+		return nil, trace.Wrap(fmt.Errorf(resp.errorStrings[int(l)]))
+	}
+
+	// Couldn't write to result since there is not enough memory allocated
+	if int(l) > len(b) {
+		res = (*C.uchar)(C.CBytes(make([]byte, l)))
+		l = C.process_access(ptr, res, C.size_t(len(b)), C.size_t(l))
+		if int(l) < 0 {
+			return nil, trace.Wrap(fmt.Errorf(resp.errorStrings[int(l)]))
+		}
+	}
+	C.deallocate_rust_buffer(res, C.size_t(l))
+	defer C.free(unsafe.Pointer(res))
+
+	err = proto.Unmarshal(C.GoBytes(unsafe.Pointer(res), C.int(l)), &resp.Accesses)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp.Accesses = cleanOutput(resp.Accesses, resp.reverseMappings)
+	resp.Accesses = filterByLogin(resp.Accesses, resp.reverseMappings, req.Login)
+	return &resp, nil
+}
+
+func NewNodeAccessResponse() NodeAccessResponse {
+	resp := NodeAccessResponse{Facts{}, Facts{}, make(map[string]uint32), make(map[uint32]string), map[int]string{
+		-1: "Invalid input.",
+		-2: "Invalid output generation.",
```

Comment:
Instead of using this protocol for mapping codes to errors, can an error message just be returned from Rust? Also, these errors are very vague, i.e. as a user if I would receive any of these, I'd have no idea what to do.

Reply:
I was debating whether to have the rust function return the byte buffer or just have a more c style function where I return a status code and have a pointer result that is modified. In this case, I wouldn't want to decode a byte buffer (which could be a string or an encoded protobuf) that doesn't match the protobuf struct since the results could be unpredictable if I just returned the byte buffer.

Reply:
One form of doing this I see somewhat often is to either have a status code and a buffer return or return a pointer where the last bit is used to tell if it was an error or not. On error the actual pointer is decoded and it points to a string buffer containing the Rust error in string format via `ToString`.

Reply:
That gives room to easily use existing Rust errors and use the rich templating and data you already have to give good errors. Sometimes a code may also be needed somehow to translate it to the appropriate error type in Go.

Reply:
This seems like the best way to do it for our use case. I'll take a look into this!

Reply:
I've decided to go with a Status struct that will wrap the error flag along with the output length.